<a href="https://colab.research.google.com/github/ZozolNat/teor_veroiatnosti-i-stat/blob/main/%D0%B43_9_%D1%81%D1%82%D0%B0%D1%82_%D0%B8_%D0%B2%D0%B5%D1%80%D0%BE%D1%8F%D1%82_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#1)Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks): zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110], ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]. Используя математические операции, посчитать коэффициенты линейной регрессии, приняв за X заработную плату (то есть, zp - признак), а за y - значения скорингового балла (то есть, ks - целевая переменная). Произвести расчет как с использованием intercept, так и без.


описание и решение

Без использования intercept по математической формуле:
beta = (Σ(x_i * y_i)) / (Σ(x_i^2))

***Коэффициент линейной регрессии без intercept (beta_without_intercept) представляет собой наклон прямой, проходящей через начало координат (0, 0). Он показывает, насколько изменяется зависимая переменная при изменении независимой переменной.

далее пишем в Пайтон

In [1]:
import numpy as np

# дано по условию, тут так же можно zp присвоить Х, а ks присвоить Y
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

# Без использования intercept
beta_without_intercept = np.sum(zp * ks) / np.sum(zp**2)



print("Коэффициент линейной регрессии без intercept:", beta_without_intercept)



Коэффициент линейной регрессии без intercept: 5.889820420132689


С использованием intercept:
beta = (Σ((x_i - mean(x)) * (y_i - mean(y)))) / (Σ((x_i - mean(x))^2))

***Коэффициент линейной регрессии с intercept (beta_with_intercept) учитывает не только наклон прямой, но и смещение (сдвиг) относительно начала координат. Он представляет собой наклон прямой, проходящей через наилучшую подгонку точек данных.

ищем средние значения X (заработная плата) и Y (скоринговый балл): mean_x, mean_y. Далее находим разность между каждым значением X и средним значением X: X - mean_x и разность между каждым значением Y и средним значением Y:Y - mean_y.
** как в задачах ДЗ ранее. Потом сумму произведений разностей. Делим на сумму квадратов разностей.


In [2]:
import numpy as np

zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

# С использованием intercept
mean_zp = np.mean(zp)
mean_ks = np.mean(ks)
beta_with_intercept = np.sum((zp - mean_zp) * (ks - mean_ks)) / np.sum((zp - mean_zp)**2)

print("Коэффициент линейной регрессии с intercept:", beta_with_intercept)

# считаем в пайтон с учетом указанной формулы

Коэффициент линейной регрессии с intercept: 2.6205388824027653


#2) Посчитать коэффициент линейной регрессии при заработной плате (zp), используя градиентный спуск (без intercept). ( продолжение задачи предыдущей)


ОПИСАНИЕ И РЕШЕНИЕ 

Берем тут начальное значение коэффициента регрессии (beta) равным нулю. далее надо установить скорость обучения (learning_rate), которая определяет шаг, с которым обновляются значения коэффициента (берем тут 0.0001 , хотя могут быть и другие). n: количество наблюдений (размер выборки), в данном случае zp.  Указываем количество итераций  ( Число повторений процесса обновления коэффициентов регрессии на каждой итерации. **Обычно начинают с небольшого числа итераций и постепенно увеличивают его до тех пор, пока не достигнут нужные результаты или пока изменения в коэффициентах станут незначительными. ), в данном случае берем 1000, однако может быть другое.

математическая формула без intercept выглядит так beta = beta - learning_rate * (-2/n) * sum((zp_i * (ks_i - beta * zp_i)) for i in range(n)), (For i in range(n) выполняет итерацию по индексам элементов списка zp и ks, где i принимает значения от 0 до n-1, и выполняет вычисления внутри цикла для каждого значения i.)


пишем в пайтоне


In [4]:
import numpy as np

# Заданные данные
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

# Градиентный спуск без intercept
def gradient_descent(zp, ks, learning_rate, num_iterations):
    n = len(zp)
    beta = 0  # Начальное значение коэффициента регрессии

    for _ in range(num_iterations):
        gradient = (-2/n) * np.dot(zp, ks - np.dot(zp, beta))  # Градиентная часть формулы
        beta -= learning_rate * gradient  # Обновление коэффициента

    return beta

# Параметры градиентного спуска
learning_rate = 0.0001
num_iterations = 1000

# Расчет коэффициента линейной регрессии
beta = gradient_descent(zp, ks, learning_rate, num_iterations)

print(f"Коэффициент линейной регрессии без intercept: {beta}")



Коэффициент линейной регрессии без intercept: -1.3378370976517613e+245


# 3) Произвести вычисления как в пункте 2, но с вычислением intercept. Учесть, что изменение коэффициентов должно производиться на каждом шаге одновременно (то есть изменение одного коэффициента не должно влиять на изменение другого во время одной итерации).


Описание и решение. СМОТРЕТЬ ЗАДАЧУ ВЫШЕ.

Для того чтоб посчитать коэффициентов линейной регрессии с использованием градиентного спуска и учетом intercept, 

А) сначала необходимо вводить дополнительный признак (B1), состоящий из единиц, чтобы представить intercept в виде отдельного коэффициента. * составить матрицу , см лекцию 9 

(X = [[1, 35],
     [1, 45],
     [1, 190],
     [1, 200],
     [1, 40],
     [1, 70],
     [1, 54],
     [1, 150],
     [1, 120],
     [1, 110]])

В) далее нициализируем начальное значения коэффициентов beta: beta = [0, 0]

Г) Установить параметры градиентного спуска, включая шаг обучения learning_rate и количество итераций n_iterations. ( как впред задаче) 0.0001 и 1000

Д) На каждой итерации градиентного спуска обновить значения коэффициентов beta:



for iteration in range(n_iterations):

 gradient = (-2/n) * np.dot(X.T, (ks - np.dot(X, beta)))

 beta = beta - learning_rate * gradient


Е) ищем значения коэффициентов линейной регрессии:

intercept = beta[0]

slope = beta[1]

*** При использовании градиентного спуска и c intercept надо искать два коэффициента: slope (наклон) и intercept (смещение).





In [5]:
import numpy as np

# Исходные данные
zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110]
ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]

# Добавляем столбец единиц к вектору признаков
X = np.column_stack((np.ones(len(zp)), zp))

# Инициализируем начальные значения коэффициентов
beta = np.zeros(X.shape[1])

# Параметры градиентного спуска
learning_rate = 0.0001
n_iterations = 1000

# Градиентный спуск
for iteration in range(n_iterations):
    gradient = (-2/len(zp)) * np.dot(X.T, (ks - np.dot(X, beta)))
    beta = beta - learning_rate * gradient

# Получаем значения коэффициентов
intercept = beta[0]
slope = beta[1]

# Выводим результаты
print("Intercept (интерсепт):", intercept)
print("Slope (наклон):", slope)



Intercept (интерсепт): -1.072057860005864e+243
Slope (наклон): -1.4565128585234624e+245


***Интерсепт (Intercept) в контексте линейной регрессии представляет собой значение, которое указывает на ожидаемое значение зависимой переменной, когда независимая переменная равна нулю. В данном случае, значение интерсепта -1.072 означает, что при заработной плате (zp) равной нулю, ожидаемое значение скорингового балла (ks) составляет -1.072.

Наклон (Slope) представляет собой значение, которое показывает, насколько изменяется зависимая переменная при изменении на единицу независимой переменной. В данном случае, значение наклона -1.456 указывает на то, что при увеличении заработной платы на одну единицу, ожидается снижение скорингового балла на 1.456.